osu.Server.Queues.ScoreStatisticsProcessor/Processors/MedalAwarders/StarRatingMedalAwarder.cs

```diff
@@ -0,0 +1,173 @@
+    {
+        public bool RunOnFailedScores => false;
+
+        public IEnumerable<Medal> Check(IEnumerable<Medal> medals, MedalAwarderContext context)
+        {
+            return checkAsync(medals, context).Result;
+        }
+
+        private async Task<IEnumerable<Medal>> checkAsync(IEnumerable<Medal> medals, MedalAwarderContext context)
+        {
+            List<Medal> earnedMedals = new List<Medal>();
+
+            Ruleset ruleset = LegacyRulesetHelper.GetRulesetFromLegacyId(context.Score.RulesetID);
+            Mod[] mods = context.Score.Mods.Select(m => m.ToMod(ruleset)).ToArray();
+
+            // Ensure the score isn't using any difficulty reducing mods
+            if (mods.Any(MedalHelpers.IsDifficultyReductionMod))
+                return Enumerable.Empty<Medal>();
+
+            // On mania, these medals cannot be triggered with key mods and Dual Stages
+            if (context.Score.RulesetID == 3 && mods.Any(isManiaDisallowedMod))
+                return Enumerable.Empty<Medal>();
+
+            try
+            {
+                Beatmap? beatmap = await context.BeatmapStore.GetBeatmapAsync((uint)context.Score.BeatmapID, context.Connection, context.Transaction);
+                if (beatmap == null)
+                    return Enumerable.Empty<Medal>();
+
+                // Make sure the map isn't Qualified or Loved, as those maps may occasionally have SR-breaking/aspire aspects
+                if (beatmap.approved == BeatmapOnlineStatus.Qualified || beatmap.approved == BeatmapOnlineStatus.Loved)
+                    return Enumerable.Empty<Medal>();
+
+                // Get map star rating (including mods)
+                APIBeatmap apiBeatmap = beatmap.ToAPIBeatmap();
+                DifficultyAttributes? difficultyAttributes = await context.BeatmapStore.GetDifficultyAttributesAsync(apiBeatmap, ruleset, mods, context.Connection, context.Transaction);
+
+                if (difficultyAttributes == null)
+                    return Enumerable.Empty<Medal>();
+
+                // Award pass medals
+                foreach (var medal in medals)
+                {
+                    if (checkMedalPass(context.Score, medal, difficultyAttributes.StarRating))
+                        earnedMedals.Add(medal);
+                }
+
+                // Check for FC and award FC medals if necessary
+                if (context.Score.MaxCombo == context.Score.MaximumStatistics.Where(kvp => kvp.Key.AffectsCombo()).Sum(kvp => kvp.Value))
+                {
+                    foreach (var medal in medals)
+                    {
+                        if (checkMedalFc(context.Score, medal, difficultyAttributes.StarRating))
+                            earnedMedals.Add(medal);
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                await Console.Error.WriteLineAsync($"{context.Score.ID} failed in StarRatingMedalAwarder with: {ex}");
```

Comment:
Do we really want this? It stops the queue item from throwing itself, triggering retries and potentially alerting from queue processor restarts.

I bring this up because it's also stopping us from using `yield return`, which simplifies code considerably.

@bdach

Reply:
Difficult to anticipate what the failure modes here are going to be but you're right, probably better to be louder about it.